### PR TITLE
[#3709] Ensure all data is read from socket when EPOLLRDUP is received

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

When EPOLLRDHUP is received we need to try to read at least one time to ensure
that we read all pending data from the socket. Otherwise we may loose data.

Modifications:

- Ensure we read all data from socket
- Ensure file descriptor is closed on doClose() even if doDeregister() throws an Exception.
- Only handle either EPOLLRDHUP or EPOLLIN as only one is needed to detect connection reset.

Result:

No more data loss on connection reset.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin please check

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 1]
```diff
@@ -304,14 +304,15 @@ private void processReady(EpollEventArray events, int ready) {
...
-
+                if (ch != null) {
                     AbstractEpollUnsafe unsafe = (AbstractEpollUnsafe) ch.unsafe();
 
-                    // We need to check if the channel is still open before try to trigger the
-                    // callbacks.
-                    // See https://github.com/netty/netty/issues/3443
-                    if ((ev & Native.EPOLLRDHUP) != 0 && ch.isOpen()) {
+                    // First check if EPOLLIN was set, in this case we not need to check for
+                    // EPOLLRDHUP as with EPOLLIN already handled the connection-reset case.
```
Comment (trustin) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java]: we not need -> we do not need

I don't understand the 'as with EPOLLIN...' clause.  Could you rephrase or fix it?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java [hunk 2]
```diff
@@ -765,11 +765,14 @@ private boolean doFinishConnect() throws Exception {
 
         @Override
         void epollRdHupReady() {
-            // Just call closeOnRead(). There is no need to trigger a read as this
-            // will result in an IOException anyway.
-            //
-            // See https://github.com/netty/netty/issues/3539
-            closeOnRead(pipeline());
+            if (isActive()) {
+                // If it is still active we need to call epollInReady as otherwise we may miss to
```
Comment (trustin) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java]: Missing ',' (comma) between 'active' and 'we'

Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java]: fixed

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 1 above)
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java]: addressed... Please check if this is better now :)

**Comment by trustin** [comment]
LGTM. Please cherry-pick!

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (59209f16bfd736f3d1dc8e7da68cb50a5d74c24f), 4.1 (59209f16bfd736f3d1dc8e7da68cb50a5d74c24f) and master (08e4b076be6402a4f98a2e082b1a292756c570ff)
